c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -485,51 +502,170 @@ impl<'a, 'tcx> ExprRewriteVisitor<'a, 'tcx> {
+        // on which we can safely adjust `Quantity`.
+        while from.own != to.own {
+            match self.cast_ownership_one_step(from, to, true) {
+                Some(new_own) => {
+                    from.own = new_own;
+                }
+                None => break,
+            }
         }
 
-        if matches!(
-            from.qty,
-            Quantity::Array | Quantity::OffsetPtr | Quantity::Slice
-        ) && to.qty == Quantity::Single
-        {
-            let mutbl = match from.own {
-                Ownership::Imm => Some(false),
+        // Safe casts that change `Quantity`.
+        while from.qty != to.qty {
+            // Mutability of `from`.  `None` here means that safe `Quantity` conversions aren't
+            // possible given `from`'s `Ownership`.  For example, we can't convert `Box<[T]>` to
+            // `Box<T>`.
+            let opt_mutbl = match from.own {
+                // Note that `Cell` + `Slice` is `&[Cell<T>]`, not `&Cell<[T]>`, so it can be
+                // handled like any other `&[_]`.
+                Ownership::Imm | Ownership::Cell => Some(false),
                 Ownership::Mut => Some(true),
-                // Note that `Cell` + `Slice` is `&[Cell<T>]`, not `&Cell<[T]>`.
-                Ownership::Cell => Some(false),
                 _ => None,
             };
-            if let Some(mutbl) = mutbl {
-                self.emit(RewriteKind::SliceFirst { mutbl });
-                from.qty = Quantity::Single;
-            }
-        }
+            match (from.qty, to.qty) {
+                (Quantity::Array, _) => {
+                    // `Array` goes only to `Slice` directly.  All other `Array` conversions go
+                    // through `Slice` first.
+                    error!("NYI: cast Array to {:?}", to.qty);
+                    from.qty = Quantity::Slice;
+                }
+                // Bidirectional conversions between `Slice` and `OffsetPtr`.
+                (Quantity::Slice, Quantity::OffsetPtr) => {
+                    let _rw = match opt_mutbl {
+                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
+                        Some(_mutbl) => (),
+                        None => break,
+                    };
+                    // self.emit(rw);
+                    from.qty = Quantity::OffsetPtr;
+                }
+                (Quantity::OffsetPtr, Quantity::Slice) => {
+                    let _rw = match opt_mutbl {
+                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
+                        Some(_mutbl) => (),
+                        None => break,
+                    };
+                    // self.emit(rw);
+                    from.qty = Quantity::Slice;
+                }
+                // `Slice` and `OffsetPtr` convert to `Single` the same way.
+                // TODO: may be better to use `slice.as_ptr()` to avoid panic on 0-length inputs
```

Comment:
```suggestion
                // TODO: We should use `slice.as_ptr()` to avoid panic on 0-length inputs
```

Can we just do this now?  What's the idea behind doing it this way (`&a[0]`) at all in the first place?

Reply:
`&x[0]` goes from `&[T]` to `&T`; `x.as_ptr()` goes from `&[T]` to `*const T`.  The latter doesn't fit well into the current design because it changes ownership and quantity at the same time, though I think we could add it as a special case if it's needed at some point.

Reply:
I'm curious, in what cases is this done where `&x[0]` is not used as a slice?